<a href="https://colab.research.google.com/github/ZiY1/CSC-44800-Artificial-Intelligence/blob/main/Assignment_03/Ziyi_Huang_AS03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/ZiY1/CSC-44800-Artificial-Intelligence/blob/main/Assignment_03/Ziyi_Huang_AS03.ipynb#scrollTo=oEm8VVoTb5dn[link text](https://)" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Imports section

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, RepeatedKFold, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [ ]:
# Using pandas load the dataset (load remotely, not locally)

df_science = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")

In [ ]:
# Output the first 15 rows of the data

df_science.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)

df_science.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)

# label (y)

y = df_science.iloc[:, 2].values

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)

# features (X)

X = df_science.iloc[:, df_science.columns != 'Size nm^3'].values

In [ ]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.9, test_size=0.1, random_state=1, shuffle = True)

## Part 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set

reg = LinearRegression().fit(X_train, y_train)

In [ ]:
# Create a sample datapoint and predict the output of that sample with the trained model

# Ler's create a sample point as Temperature °C: 28; Mols KCL: 600.

X_sample_point = np.array([28, 600]).reshape(1, -1) # Reshape to a 2D array

print('Sample Point: ', X_sample_point[0][0], ',', X_sample_point[0][1],)

predict = reg.predict(X_sample_point)

print('Prediction: ', predict[0])

Sample Point:  28 , 600
Prediction:  233100.8901880661


In [ ]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

train_accuray = reg.score(X_train, y_train)
test_accuray = reg.score(X_test, y_test)

print('Training Accuracy: ', '{:0.5f}'.format(train_accuray))
print('Testing Accuracy: ', '{:0.5f}'.format(test_accuray))

Training Accuracy:  0.86088
Testing Accuracy:  0.85809


## Report on the score of the model

### 1. what is the score function?
In simply words, score function is used to compute the models' prediction accuracy score, where by deafult it is calulated by correct predictions / all prediction.
### 2. what does the score mean in our model? 
- We get the training accuracy by calling reg.score(X_train, y_train), where our model make the predictions on the traning dataset that we have splitted previously. We got the result of 0.86088, which means that in the all prediction that our model made in training dataset, 86.088% of them are correct.
- We get the testing accuracy by calling reg.score(X_test, y_test), where our model make the predictions on the testing dataset at this time. We got the result of 0.85809, which means that in the all prediction that our model made in testing dataset, 85.809% of them are correct.
- It is interesting to notice that our training accuracy is slightly better than testing accuracy. This is because that our model has seen the training dataset before, and it trained on that.

In [ ]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

reg_coef = reg.coef_ # Extract the coefficents
reg_intercept = reg.intercept_ # Extract the intercept

print('Intercept:')
print('b0:', '{:0.5f}'.format(reg_intercept))
print('Coefficient:')
print('b1: ', '{:0.5f}'.format(reg_coef[0]))
print('b2: ', '{:0.5f}'.format(reg_coef[1]))

Intercept:
b0: -413045.20674
Coefficient:
b1:  861.67892
b2:  1036.69848


## Equation for your h(x)

### Linear Regresion Equation

## $ h(x) = b_{0} + b_{1}\cdot x_{1} + b_{2}\cdot x_{2} + \ldots + b_{n}\cdot x_{n} $
where $b_0$ is the intercept, $b_1$, $b_2$ are coefficients, and n is No. of observatioons 

## $ h(x) = -413045.20674 + 861.67892 \cdot x_{1} + 1036.69848 \cdot x_{2} $

## Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

rkf  = RepeatedKFold(n_splits = 10, n_repeats=100, random_state = 1)
scores = cross_val_score(reg, X, y, cv=rkf)
#print(scores)
print('Accuracy: %.5f (%.5f)' % (scores.mean(), scores.std()))

Accuracy: 0.85674 (0.02097)


## Report on their finding and their significance

The cross-validation is used to evaluate the estimator's performance. More specifically, it is used to estimate the skill of our model on unseen data. Sometimes overfitting occurs when our model performs well on training data but does not generalize well to unseen test data, and cross validation can help use identify the overfitting problem. In this case, we can see that we cross validation gives us the mean score of 0.85674 with a standard deviation of 0.02097, so that we can confirm that our model does not have a overfitting problem because it generalizes well to unseen test data, recall that we have a testing accuracy of 0.85089

## Part 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

poly = PolynomialFeatures(2).fit(X_train, y_train)

# Augment the data

X_aug = poly.fit_transform(X)

# Split the dataset as before

X_train_aug, X_test_aug, y_train, y_test = train_test_split(X_aug, y, train_size = 0.9, test_size=0.1, random_state=1, shuffle = True)

# Perform a new regression

poly_reg = LinearRegression().fit(X_train_aug, y_train)

# Score of the model

train_accuray = poly_reg.score(X_train_aug, y_train)
test_accuray = poly_reg.score(X_test_aug, y_test)

print('Training Accuracy: ', train_accuray)
print('Testing Accuracy: ', test_accuray)

Training Accuracy:  1.0
Testing Accuracy:  1.0


## Report on the metrics
Polynomial regression fits the data to more complex polynomial function.By using the Polynomial Regression in our model, we obtained an traing acuuracy of 1.0 and testing acurracy of 1.0, which is a full score, and it is better than the linear regression model. The reason is that polynomial fits a wide range of curvature, which is more flexiable than linear.

In [ ]:
# output the resultant equation as you did in Part 3.

# The polunomial regression equation

poly_reg_coef = poly_reg.coef_ # Extract the coefficents
poly_reg_intercept = poly_reg.intercept_ # Extract the intercept
print('Intercept:')
print('b0:', '{:0.5f}'.format(poly_reg_intercept))
print('Coefficient:')
for i in range(len(poly_reg_coef)):
    print('b', i, ':', '{:0.5f}'.format(poly_reg_coef[i]))

Intercept:
b0: 0.00001
Coefficient:
b 0 : 0.00000
b 1 : 12.00000
b 2 : -0.00000
b 3 : -0.00000
b 4 : 2.00000
b 5 : 0.02857


## Polynomial Regresion Equation

##  $h(x) = 0.00001 + 0 \cdot1 + 12\cdot x_1 - 0\cdot x_2 - 0 \cdot x_1^2 + 2\cdot x_1 \cdot x_2 + 0.02857 \cdot x_2^2 $ 

## $ = 0.00001 + 12\cdot x_1 + 2\cdot x_1 \cdot x_2 + 0.02857 \cdot x_2^2 $ 